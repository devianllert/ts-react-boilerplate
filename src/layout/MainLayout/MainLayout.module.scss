@import '../../design/design-system';

.header, .footer {
  width: 100%;
  height: 60px;

  display: flex;
  justify-content: space-between;
  align-items: center;

  padding: 0 $space-24;

  @include mediaMax('phone') {
    padding: 0 $space-8;
  }
}

.header {
  box-shadow: 0 0 6px 0 rgba(#000, 0.5);
}

.footer {
  border-top: 1px solid rgba(#000, 0.25);
}

.footerLogo {
  font-size: $font-size;
  font-weight: $font-weight-light;

}

.logo {
  font-size: $font-size-h3;
  font-weight: $font-weight-h3;

  text-decoration: none;

  @include mediaMax('phone') {
    font-size: $font-size-h4;
    font-weight: $font-weight-h4;
  }
}

.menu {
  display: flex;

  height: 100%;

  margin: 0;
  padding: 0;

  list-style-type: none;

  @include mediaMax('phone') {
    display: none;
  }
}

.menuItem {
  display: flex;
  
  align-items: center;
  justify-content: center;

  height: 100%;

  font-size: $font-size;
  font-weight: $font-weight-regular;

  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;

  &:hover {
    background-color: var(--brand-ghost);
  }

  a {
    display: flex;
  
    align-items: center;
    justify-content: center;
  
    height: 100%;

    padding: 0 $space-16;

    &.active {
      background-color: var(--brand-ghost);
    }
  }
}

.content {
  display: flex;
  flex-direction: column;
  flex: 1;

  min-height: calc(100vh - 120px);

  padding: $space-24;

  @include mediaMax('phone') {
    padding: $space-8;
  }
}