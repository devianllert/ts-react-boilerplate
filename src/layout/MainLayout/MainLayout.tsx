import React, { ReactChild, ReactElement } from 'react';
import { Link, NavLink } from 'react-router-dom';
import { useTranslation } from 'react-i18next';

import LocaleToggler from '../../components/LocaleToggler';
import ThemeToggler from '../../components/ThemeToggler';

import styles from './MainLayout.module.scss';

interface Props {
  children: ReactChild | ReactChild[];
}

const MainLayout = ({ children }: Props): ReactElement => {
  const { t } = useTranslation();

  return (
    <>
      <header className={styles.header}>
        <Link to="/" className={styles.logo}>TS BOILERPLATE</Link>

        <ul className={styles.menu}>
          <li className={styles.menuItem}>
            <NavLink exact activeClassName={styles.active} to="/">{t('layout.main.menu.home')}</NavLink>
          </li>
          <li className={styles.menuItem}>
            <NavLink activeClassName={styles.active} to="/counter">{t('layout.main.menu.counter')}</NavLink>
          </li>
          <li className={styles.menuItem}>
            <NavLink activeClassName={styles.active} to="/advice">{t('layout.main.menu.advice')}</NavLink>
          </li>
          <li className={styles.menuItem}>
            <NavLink activeClassName={styles.active} to="/repos">{t('layout.main.menu.repos')}</NavLink>
          </li>
          <li className={styles.menuItem}>
            <NavLink activeClassName={styles.active} to="/about">{t('layout.main.menu.about')}</NavLink>
          </li>
          <li className={styles.menuItem}>
            <LocaleToggler />
          </li>
          <li className={styles.menuItem}>
            <ThemeToggler />
          </li>
        </ul>

      </header>

      <main className={styles.content}>{children}</main>

      <footer className={styles.footer}>
        <span className={styles.footerLogo}>TS Boilerplate</span>
      </footer>
    </>
  );
};

export default MainLayout;
