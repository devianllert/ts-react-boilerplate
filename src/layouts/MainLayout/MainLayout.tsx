import React, { ReactNode, ReactElement } from 'react';

import * as S from './styled';

interface Props {
  children: ReactNode;
}

const MainLayout = (props: Props): ReactElement => {
  const { children } = props;

  return <S.Main>{children}</S.Main>;
};

export default MainLayout;
