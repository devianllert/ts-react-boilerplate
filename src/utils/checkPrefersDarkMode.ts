/**
 * Check browser or OS preference for dark mode
 */

const checkPrefersDarkMode = (): boolean => {
  const preferDarkQuery = '(prefers-color-scheme: dark)';

  const mediaQuery = matchMedia(preferDarkQuery);

  return mediaQuery.media === preferDarkQuery;
};

export default checkPrefersDarkMode;
