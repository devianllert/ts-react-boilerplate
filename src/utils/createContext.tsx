import React, {
  createContext,
  useState,
  useContext,
  Provider,
  Dispatch,
  SetStateAction,
  PropsWithChildren,
  ReactNode,
  Context,
} from 'react';

type UpdateType<A> = Dispatch<SetStateAction<A>>;

/**
 * A helper to create a Context and Provider with no upfront default value, and
 * without having to check for undefined all the time.
 */
export const createCtx = <A extends {} | null>(): readonly [() => A, Provider<A | undefined>] => {
  const ctx = createContext<A | undefined>(undefined);

  const useCtx = (): A => {
    const c = useContext(ctx);

    if (c === undefined) {
      throw new Error('useCtx must be inside a Provider with a value');
    }

    return c;
  };

  return [useCtx, ctx.Provider] as const;
};

export const createCtxWithUpdate = <A extends any>(
  defaultValue: A,
): readonly [
  Context<{ state: A; update: Dispatch<SetStateAction<A>> }>,
  (props: { children?: ReactNode }) => ReactNode,
] => {
  const defaultUpdate: UpdateType<typeof defaultValue> = () => defaultValue;

  const ctx = createContext({
    state: defaultValue,
    update: defaultUpdate,
  });

  const ContextProvider = (props: PropsWithChildren<{}>): ReactNode => {
    const [state, update] = useState(defaultValue);
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <ctx.Provider value={{ state, update }} {...props} />;
  };

  return [ctx, ContextProvider] as const;
};
