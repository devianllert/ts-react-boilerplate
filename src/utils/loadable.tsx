/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable react/jsx-props-no-spreading */

import React, {
  lazy,
  Suspense,
  ReactElement,
  ComponentType,
} from 'react';

export type ImportComponent = Promise<{ default: any }>;

const loadable = (importFunc: () => ImportComponent, { fallback = null }): ComponentType => {
  const LazyComponent: ComponentType = lazy(importFunc);

  return (props: object): ReactElement => (
    <Suspense fallback={fallback}>
      <LazyComponent {...props} />
    </Suspense>
  );
};

export default loadable;
