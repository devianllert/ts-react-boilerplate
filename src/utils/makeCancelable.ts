/* eslint-disable prefer-promise-reject-errors */
/* https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html */
export interface CancellablePromise<T> {
  promise: Promise<T>;
  cancel(): void;
}

const makeCancelable = <T>(promise: Promise<T>): CancellablePromise<T> => {
  let hasCanceled = false;

  const wrappedPromise = new Promise<T>((resolve, reject) => {
    promise.then(
      (val) => (hasCanceled ? reject({ isCanceled: true }) : resolve(val)),
      (error) => (hasCanceled ? reject({ isCanceled: true }) : reject(error)),
    );
  });

  return {
    promise: wrappedPromise,
    cancel(): void {
      hasCanceled = true;
    },
  };
};

export default makeCancelable;
