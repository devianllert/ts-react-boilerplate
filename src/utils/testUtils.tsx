/* eslint-disable import/prefer-default-export */
import React, { ReactElement } from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { i18n as i18Next } from 'i18next';
import { I18nextProvider } from 'react-i18next';
import { render, RenderResult } from '@testing-library/react'; // eslint-disable-line

import configureStore, { EnhancedStore } from '../core/configureStore';
import i18nInstance from '../core/i18n';

export interface RenderProvidersOptions {
  initialState?: {};
  route?: string;
  routerHistory?: MemoryHistory;
  store?: EnhancedStore;
  i18n?: i18Next;
}

export interface RenderProvidersResult extends RenderResult {
  routerHistory: MemoryHistory;
  store: EnhancedStore;
  i18n: i18Next;
}

export type RenderProviders = (ui: ReactElement, options?: RenderProvidersOptions) => RenderProvidersResult;

const renderWithProviders: RenderProviders = (
  ui,
  {
    initialState = {},
    route = '/',
    routerHistory = createMemoryHistory({ initialEntries: [route] }),
    store = configureStore(initialState),
    i18n = i18nInstance,
  }: RenderProvidersOptions = {},
): RenderProvidersResult => ({
  ...render(
    <Provider store={store}>
      <I18nextProvider i18n={i18n}>
        <Router history={routerHistory}>{ui}</Router>
      </I18nextProvider>
    </Provider>,
  ),
  // adding `history` to the returned utilities to allow us
  // to reference it in our tests (just try to avoid using
  // this to test implementation details).
  store,
  routerHistory,
  i18n,
});

export {
  renderWithProviders,
};
