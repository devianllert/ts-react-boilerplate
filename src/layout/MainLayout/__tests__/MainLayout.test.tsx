import React, { ReactChild } from 'react';
import { cleanup } from '@testing-library/react';

import '@testing-library/jest-dom/extend-expect';

import { renderWithRouter } from '../../../utils/testUtils';

import MainLayout from '../MainLayout';

describe('<MainLayout />', (): void => {
  afterEach(cleanup);

  const children: ReactChild = 'MainLayout';

  it('should render a children', (): void => {
    const { getByText } = renderWithRouter(<MainLayout>{children}</MainLayout>);

    expect(getByText(children)).toHaveTextContent(children);
  });

  it('should render and match the snapshot', (): void => {
    const mainLayout = renderWithRouter(<MainLayout>{children}</MainLayout>);

    expect(mainLayout.container).toMatchSnapshot();
  });
});
