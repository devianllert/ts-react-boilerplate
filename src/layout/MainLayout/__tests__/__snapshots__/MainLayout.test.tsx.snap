// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainLayout /> should render and match the snapshot 1`] = `
<div>
  <header
    class="header"
  >
    <a
      class="logo"
      href="/"
    >
      TS BOILERPLATE
    </a>
    <ul
      class="menu"
    >
      <li
        class="menuItem"
      >
        <a
          aria-current="page"
          class="active"
          href="/"
        >
          layout.main.menu.home
        </a>
      </li>
      <li
        class="menuItem"
      >
        <a
          href="/counter"
        >
          layout.main.menu.counter
        </a>
      </li>
      <li
        class="menuItem"
      >
        <a
          href="/advice"
        >
          layout.main.menu.advice
        </a>
      </li>
      <li
        class="menuItem"
      >
        <a
          href="/about"
        >
          layout.main.menu.about
        </a>
      </li>
      <li
        class="menuItem"
      >
        <div
          class="toggler"
          role="menu"
          tabindex="0"
        >
          <span
            class="title"
          />
          <ul
            class="menu"
          >
            <li
              class="menuItem"
              role="menuitem"
              tabindex="0"
            >
              en
            </li>
            <li
              class="menuItem"
              role="menuitem"
              tabindex="0"
            >
              ru
            </li>
          </ul>
        </div>
      </li>
      <li
        class="menuItem"
      >
        <div
          class="toggler"
          role="button"
          tabindex="0"
        >
          <span
            class="title"
          >
            dark
          </span>
        </div>
      </li>
    </ul>
  </header>
  <main
    class="content"
  >
    MainLayout
  </main>
  <footer
    class="footer"
  >
    <span
      class="footerLogo"
    >
      TS Boilerplate
    </span>
  </footer>
</div>
`;
